src/darwin/Framework/CHIP/MTRBaseDevice.mm

```diff
@@ -1356,29 +1356,36 @@ - (void)subscribeToAttributePaths:(NSArray<MTRAttributeRequestPath *> * _Nullabl
                        });
                    };
 
-                   auto establishedOrFailed = chip::Platform::MakeShared<BOOL>(NO);
-                   auto onFailureCb = [establishedOrFailed, queue, subscriptionEstablished, reportHandler](
-                                          const app::ConcreteAttributePath * attributePath,
+                   auto onFailureCb = [queue, reportHandler](const app::ConcreteAttributePath * attributePath,
                                           const app::ConcreteEventPath * eventPath, CHIP_ERROR error) {
-                       // TODO, Requires additional logic if attributePath or eventPath is not null
-                       if (!(*establishedOrFailed)) {
-                           *establishedOrFailed = YES;
-                           if (subscriptionEstablished) {
-                               dispatch_async(queue, subscriptionEstablished);
-                           }
-                       }
-                       if (reportHandler) {
+                       if (attributePath != nullptr) {
+                           ConcreteAttributePath pathCopy(*attributePath);
+                           dispatch_async(queue, ^{
+                               reportHandler(@[ @ {
+                                   MTRAttributePathKey : [[MTRAttributePath alloc] initWithPath:pathCopy],
+                                   MTRErrorKey : [MTRError errorForCHIPErrorCode:error]
+                               } ],
+                                   nil);
+                           });
+                       } else if (eventPath != nullptr) {
+                           ConcreteEventPath pathCopy(*eventPath);
                            dispatch_async(queue, ^{
-                               reportHandler(nil, [MTRError errorForCHIPErrorCode:error]);
+                               reportHandler(@[ @ {
+                                   MTRAttributePathKey : [[MTREventPath alloc] initWithPath:pathCopy],
+                                   MTRErrorKey : [MTRError errorForCHIPErrorCode:error]
+                               } ],
+                                   nil);
                            });
+                       } else {
+                           if (reportHandler) {
```

Comment:
Looks like you're no longer checking whether `reportHandler` is non-null elsewhere, so can this check be removed too?

Reply:
Good catch.  That argument is nonnull, so we shouldn't be checking it...